cartographer/mapping/internal/2d/ray_casting.cc

```diff
@@ -18,79 +18,75 @@
-                           std::min(begin.y(), end.y()) / kSubpixelScale);
-    const int end_y = std::max(begin.y(), end.y()) / kSubpixelScale;
+  if (begin.x() / subpixel_scale == end.x() / subpixel_scale) {
+    Eigen::Array2i current(begin.x() / subpixel_scale,
+                           std::min(begin.y(), end.y()) / subpixel_scale);
+    ray.push_back(current);
+    const int end_y = std::max(begin.y(), end.y()) / subpixel_scale;
     for (; current.y() <= end_y; ++current.y()) {
-      probability_grid->ApplyLookupTable(current, miss_table);
+      if (!ray.back().isApprox(current)) ray.push_back(current);
     }
-    return;
+    return ray;
   }
 
   const int64 dx = end.x() - begin.x();
   const int64 dy = end.y() - begin.y();
-  const int64 denominator = 2 * kSubpixelScale * dx;
+  const int64 denominator = 2 * subpixel_scale * dx;
 
   // The current full pixel coordinates. We begin at 'begin'.
-  Eigen::Array2i current = begin / kSubpixelScale;
+  Eigen::Array2i current = begin / subpixel_scale;
+  ray.push_back(current);
 
-  // To represent subpixel centers, we use a factor of 2 * 'kSubpixelScale' in
+  // To represent subpixel centers, we use a factor of 2 * 'subpixel_scale' in
   // the denominator.
-  // +-+-+-+ -- 1 = (2 * kSubpixelScale) / (2 * kSubpixelScale)
+  // +-+-+-+ -- 1 = (2 * subpixel_scale) / (2 * subpixel_scale)
   // | | | |
   // +-+-+-+
   // | | | |
-  // +-+-+-+ -- top edge of first subpixel = 2 / (2 * kSubpixelScale)
-  // | | | | -- center of first subpixel = 1 / (2 * kSubpixelScale)
-  // +-+-+-+ -- 0 = 0 / (2 * kSubpixelScale)
+  // +-+-+-+ -- top edge of first subpixel = 2 / (2 * subpixel_scale)
+  // | | | | -- center of first subpixel = 1 / (2 * subpixel_scale)
+  // +-+-+-+ -- 0 = 0 / (2 * subpixel_scale)
 
   // The center of the subpixel part of 'begin.y()' assuming the
   // 'denominator', i.e., sub_y / denominator is in (0, 1).
-  int64 sub_y = (2 * (begin.y() % kSubpixelScale) + 1) * dx;
+  int64 sub_y = (2 * (begin.y() % subpixel_scale) + 1) * dx;
 
   // The distance from the from 'begin' to the right pixel border, to be divided
-  // by 2 * 'kSubpixelScale'.
+  // by 2 * 'subpixel_scale'.
   const int first_pixel =
-      2 * kSubpixelScale - 2 * (begin.x() % kSubpixelScale) - 1;
+      2 * subpixel_scale - 2 * (begin.x() % subpixel_scale) - 1;
   // The same from the left pixel border to 'end'.
-  const int last_pixel = 2 * (end.x() % kSubpixelScale) + 1;
+  const int last_pixel = 2 * (end.x() % subpixel_scale) + 1;
 
   // The full pixel x coordinate of 'end'.
-  const int end_x = std::max(begin.x(), end.x()) / kSubpixelScale;
+  const int end_x = std::max(begin.x(), end.x()) / subpixel_scale;
 
   // Move from 'begin' to the next pixel border to the right.
   sub_y += dy * first_pixel;
   if (dy > 0) {
     while (true) {
-      probability_grid->ApplyLookupTable(current, miss_table);
+      if (!ray.back().isApprox(current)) ray.push_back(current);
```

Comment:
is it guaranteed that the same pixel can only be at ray.back()?
Don't use isApprox() to compare integer vectors

Reply:
> is it guaranteed that the same pixel can only be at ray.back()?
Yes.